### PR TITLE
Fix race conditions

### PR DESCRIPTION
**Pull request by ashanbrown**
These may be breaking the refreshing_wiltfilter integration tests and they are certainly being picked up by the race detector our integration tests.

### CODE REVIEW

File: gocron.go [hunk 1]
```diff
@@ -112,10 +114,11 @@ func (j *Job) run() ([]reflect.Value, error) {
 			}
 			in[k] = reflect.ValueOf(param)
 		}
+		j.mu.Unlock()
```
Comment (arun251) [inline gocron.go:123]: Why is the `Unlock` here and not earlier?

Comment (ashanbrown) [reply gocron.go:123]: I believe that params itself is a shared map that needs to be protected until we copy everything out of it in into `in`.

File: gocron.go [hunk 2]
```diff
@@ -391,22 +394,16 @@ type Scheduler struct {
...
 
 func (s *Scheduler) Len() int {
-	s.mu.Lock()
 	l := len(s.jobs)
-	s.mu.Unlock()
 	return l
 }
 
 func (s *Scheduler) Swap(i, j int) {
-	s.mu.Lock()
```
Comment (ashanbrown) [inline gocron.go:401]: I think we need to do a lock around the entire sort to ensure that that the slice is stable during the sort.

File: gocron.go [hunk 3]
```diff
@@ -454,13 +455,14 @@ func (s *Scheduler) Err() error {
 
 // RunPending runs all the jobs that are scheduled to run.
 func (s *Scheduler) RunPending() error {
+	s.mu.Lock()
```
Comment (ashanbrown) [inline gocron.go]: Just because the line below is a boolean, does not make access to it threadsafe.

**Comment by ashanbrown** [comment]
I've tried to fix some obvious stuff here, but I'm not sure yet how to approach the cases where we iterate through a list of jobs.

File: gocron.go (hunk 2 above)
Comment (ashanbrown) [reply gocron.go:401]: So I've put a lock around anything that does a sort.

**Review by arun251 — DISMISSED** [review]
👍

**Review by zsiddique — APPROVED** [review]

**Review by arun251 — APPROVED** [review]
👍

**Pull request merged**